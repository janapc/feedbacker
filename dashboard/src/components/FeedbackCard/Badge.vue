<template>
  <span id="bagde-component" :class="`bg-${classColor}`" class="p-2 text-xs font-black text-white uppercase rounded-full">{{
    label
  }}</span>
</template>

<script>
import { computed } from 'vue';

export default {
  props: {
    type: { type: String, required: true }
  },
  setup (props) {
    const label = computed(() => {
      if (props.type === 'ISSUE') {
        return 'problema';
      }

      if (props.type === 'IDEA') {
        return 'idÃ©ia';
      }

      return 'outros';
    });

    const classColor = computed(() => {
      if (props.type === 'ISSUE') {
        return 'brand-danger';
      }

      if (props.type === 'IDEA') {
        return 'brand-warning';
      }

      return 'brand-graydark';
    });

    return {
      classColor,
      label
    };
  }
};
</script>
