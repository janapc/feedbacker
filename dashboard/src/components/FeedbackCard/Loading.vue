<template>
  <content-loader class="flex flex-col items-center rounded" width="100%" heigh="300px">
    <content-loader class="mt-3 rounded" width="90%" heigh="90" animation-duration="2s" />
    <content-loader class="mt-3 rounded" width="90%" heigh="90" animation-duration="2.3s" />
    <content-loader class="mt-3 rounded" width="90%" heigh="90" animation-duration="2.7s" />
  </content-loader>
</template>

<script>
import ContentLoader from '../ContentLoader/index.vue';

export default {
  components: { ContentLoader }
};
</script>
