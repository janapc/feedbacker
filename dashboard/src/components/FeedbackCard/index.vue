<template>
  <div id="feedback-card" @click="handleToggle" class="flex flex-col px-8 py-6 rounded cursor-pointer bg-brand-gray">
    <div class="flex items-center justify-between w-full mb-8">
      <badge :type="feedback.type" id="field-badge"/>

      <span class="font-regular text-brand-graydark" id="field-createdAt">
        {{ getDiffTimeBetweenCurrentDate(feedback.createdAt) }}
      </span>
    </div>

    <div class="text-lg font-medium text-gray-800" id="field-text">
      {{ feedback.text }}
    </div>

    <div
      :class="{
        animate__fadeOutUp: state.isClosing
      }"
      class="flex mt-8 animate__animated animate__fadeInUp animate__faster"
      v-if="state.isOpen"
      id="open-card"
    >
      <div class="flex flex-col w-1/2">
        <div class="flex flex-col">
          <span class="font-bold text-gray-400 uppercase select-none">Página</span>
          <span class="font-medium text-gray-700" id="field-page">{{ feedback.page }}</span>
        </div>

        <div class="flex flex-col">
          <span class="font-bold text-gray-400 uppercase select-none">Dispositivo</span>
          <span class="font-medium text-gray-700" id="field-device">{{ feedback.device }}</span>
        </div>
      </div>

      <div class="flex flex-col w-1/2">
        <div class="flex flex-col">
          <span class="font-bold text-gray-400 uppercase select-none">Usuário</span>
          <span class="font-medium text-gray-700" id="field-fingerprint">{{ feedback.fingerprint }}</span>
        </div>
      </div>
    </div>

    <div class="flex justify-end mt-8" v-if="!state.isOpen" id="close-chevron">
      <icon name="chevron-down" size="24" :color="brandColors.graydark" />
    </div>
  </div>
</template>

<script>
import { reactive } from 'vue';
import Badge from './Badge.vue';
import getDiffTimeBetweenCurrentDate from '../../utils/date';
import palette from '../../../palette';
import Icon from '../Icon/index.vue';
import wait from '../../utils/timeout';

export default {
  components: { Badge, Icon },
  props: {
    feedback: { type: Object, required: true },
    isOpened: { type: Boolean, default: false }
  },
  setup (props) {
    const state = reactive({
      isOpen: props.isOpened,
      isClosing: !props.isOpened
    });

    async function handleToggle () {
      state.isClosing = true;

      await wait(250);
      state.isOpen = !state.isOpen;
      state.isClosing = false;
    }

    return {
      handleToggle,
      getDiffTimeBetweenCurrentDate,
      brandColors: palette.brand,
      state
    };
  }
};
</script>
