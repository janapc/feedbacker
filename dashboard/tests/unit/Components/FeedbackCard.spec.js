import { shallowMount, flushPromises } from '@vue/test-utils';

import FeedbackCard from '../../../src/components/FeedbackCard/index.vue';

jest.mock('../../../src/utils/timeout');
jest.mock('../../../src/utils/date', () => () => '98 dias atrás');

describe('<FeedbackCard />', () => {
  afterEach(() => {
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it('should render feedback card correctly', async () => {
    const wrapper = shallowMount(FeedbackCard, {
      props: {
        feedback: {
          text: 'Banana Banana!',
          fingerprint: '123',
          id: '1',
          apiKey: 'kaidke-123',
          type: 'OTHER',
          device: 'Chrome 85.0, macOS 10.14',
          page: 'https://banana.com/banana',
          createdAt: 1605225600000,
        },
        isOpened: false,
      },
    });

    const bagde = wrapper.findComponent('#field-badge');
    const fieldText = wrapper.find('#field-text');
    const fieldCreatedAt = wrapper.find('#field-createdAt');
    const openCard = wrapper.find('#open-card');
    const closeChevron = wrapper.find('#close-chevron');

    expect(bagde.exists()).toBe(true);
    expect(bagde.props().type).toBe('OTHER');

    expect(fieldText.text()).toBe('Banana Banana!');
    expect(fieldCreatedAt.text()).toBe('98 dias atrás');
    expect(openCard.exists()).toBe(false);
    expect(closeChevron.isVisible()).toBe(true);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should open feedback card', async () => {
    const wrapper = shallowMount(FeedbackCard, {
      props: {
        feedback: {
          text: 'Banana Banana!',
          fingerprint: '123',
          id: '1',
          apiKey: 'kaidke-123',
          type: 'OTHER',
          device: 'Chrome 85.0, macOS 10.14',
          page: 'https://banana.com/banana',
          createdAt: 1605225600000,
        },
      },
    });

    const btnToggle = wrapper.find('#feedback-card');

    await btnToggle.trigger('click');
    await flushPromises();

    expect(wrapper.findComponent('#field-badge').exists()).toBe(true);
    expect(wrapper.findComponent('#field-badge').props().type).toBe('OTHER');

    expect(wrapper.find('#open-card').exists()).toBe(true);

    expect(wrapper.find('#field-text').text()).toBe('Banana Banana!');
    expect(wrapper.find('#field-createdAt').text()).toBe('98 dias atrás');
    expect(wrapper.find('#field-page').text()).toBe('https://banana.com/banana');
    expect(wrapper.find('#field-device').text()).toBe('Chrome 85.0, macOS 10.14');
    expect(wrapper.find('#field-fingerprint').text()).toBe('123');
    expect(wrapper.find('#close-chevron').exists()).toBe(false);
  });
});
