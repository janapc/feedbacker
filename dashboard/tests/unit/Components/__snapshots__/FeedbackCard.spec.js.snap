// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedbackCard /> should render feedback card correctly 1`] = `
<div id="feedback-card" class="flex flex-col px-8 py-6 rounded cursor-pointer bg-brand-gray">
  <div class="flex items-center justify-between w-full mb-8">
    <badge-stub type="OTHER" id="field-badge"></badge-stub><span class="font-regular text-brand-graydark" id="field-createdAt">98 dias atr√°s</span>
  </div>
  <div class="text-lg font-medium text-gray-800" id="field-text">Banana Banana!</div>
  <!--v-if-->
  <div class="flex justify-end mt-8" id="close-chevron">
    <icon-stub name="chevron-down" size="24" color="#C0BCB0"></icon-stub>
  </div>
</div>
`;
